Update stewardship guildelines and process

ksen0 opened the pull request:
Resolves #5719

This PR proposes a more structured approach to stewardship, to streamline the [existing stewardship model](https://github.com/processing/p5.js/issues/5719).

I look forward to feedback of current, past, and prospective stewards as well as contributors on any improvements - either technical or conceptual. Please feel free to add code review and comments on anything in this PR, comment in this thread!

In this PR, you will find:

* An update to the steward guidelines  that covers **what stewardship is**, **what are the areas**, and **how to become a steward**
* A structured YAML file `stewards.yml` that contains the current list of active stewards
* A GitHub Actions automation that will update the Stewards table in the repository's README every time this file is updated. (These will be synced between the `main` and `dev-2.0` branch.)

In the future, we hope to work on:

* _Technical improvements:_ developing a reusable GitHub Action rather than a script in utils which could support other projects interested in adopting a similar model
* _Governance improvements_: learning from other open source communities on how they do community care and collective decision making about software

Anyone who is interested in collectively revising on the stewardship process itself in the future, please feel free to comment accordingly! The proposed revision is based on discussion within the maintainers team, and we would be happy for input / co-creation of any future revisions.

ksen0 commented:
Currently active stewards include: @calebfoss @davepagurek @dhowe @qianqianye @ogbabydiesal @limzykenneth @perminder-17 @lukeplowden @Divyansh013 @GregStanton @holomorfo - if you are on the list and you prefer to step down, or change your area(s), please comment below! The list is made based on the existing table and on recent activity.

While preparing this proposal we (the maintainers) also saw recently active contributors who seem already taking on stewardship activities, so anyone on the below list is welcome to comment if you'd like to be a steward (you can suggest different area(s)):

* @error-four-o-four (DevOps)
* @clairep94 (p5.js-website)
* @VANSH3104 (Documentation)
* @Vaivaswat2244 (DevOps)
* @takshittt (i18n)
* @irenjie95 (i18n)
* @Darrenhqf (i18n)
* @Forchapeatl (WebGL or Shapes)

You're welcome to comment with your interest and ideas below whether or not you are in either of these lists!

lirenjie95 commented:
Could you please add me as the steward of the build process as well? I have also previously contributed to the GitHub actions. Thank you!

limzykenneth commented on contributor_docs/steward_guidelines.md:
```diff
@@ -28,6 +32,57 @@ Whether you have just joined us as a steward, are a seasoned maintainer of p5.js
+
+1. Welcoming new contributors to GitHub through **friendly comments** and helpful **code reviews**
+2. Helping to **facilitate feature discussions** and resolve technical disagreements
+3. Suporting **releases** of the p5.js software through bug reporting, bug fixing, feature completion
```
```suggestion
3. Suporting releases of the p5.js software through **supporting** bug fixing and feature completion
```

Divyansh013 replied:
"Supporting" very minor typo, in case it wasn't intended

Divyansh013 commented:
Very good initiative, happy to help!

ksen0 commented:
Thanks so much for the suggestions @limzykenneth ! In addition to your edits, I added some more clarifications to the list of steward responsibilities, and how that's distinct from contribution: 

> 1. Welcoming new contributors to GitHub through **friendly comments** and helpful **code reviews**
> 2. Helping to **facilitate feature discussions** and resolve technical disagreements. For example, this means making connections to other past discussion, or offering input from prior relevant experience. Suggesting features or working on them is contribution, rather than stewardship.
> 3. Participating in releases of the p5.js software through **supporting** bug fixing and feature completion. For example, this means guiding other contributors and reviewing their PRs. Fixing bugs is contribution, rather than stewardship.

@RandomGamingDev What do you think about the above? So, the intent here is that the stewards are focusing on community care. If your main interest is to propose and work on features, then you are welcome to continue doing that as a contributor. As a steward, practically speaking, the focus is on reviewing PRs and helping to triage bugs (ie, helping people clarify the bug, and helping other contributors fix it). Is that something you are interested in as well, and which of the areas (Math, Core, or WebGL) would be most relevant? Unlike in the previous process, the goal here is to keep the list of stewards quite focused, with 1-3 per area.

@lirenjie95 I also added a small clarification on translation stewardship. As above, it's more about reviewing or helping triage - though of course you are welcome to make translations too! Let me know what your thoughts are about this approach to stewardship on DevOps.

Vaivaswat2244 commented:
Thank you for the invitation. It would be an honor to be a steward for p5.js. I'd be happy to be a steward for DevOps category.

lirenjie95 commented on .github/workflows/stewards-update.yml:
```diff
@@ -0,0 +1,33 @@
+      - name: Run table generator
+        run: node utils/stewards-table.js
+        
+      - name: Create Pull Request
+        uses: peter-evans/create-pull-request@v5
+        with:
+          commit-message: "Update README table from stewards.yml"
+          branch: update-readme-table
+          title: 'chore: update README table from stewards.yml'
+          body: 'This PR updates the README.md table to refelct changes in stewards.yml.'
+          add: README.md
```
Adding `delete-branch` option might be helpful to manage new branches. Or we should use` GitHub's Automatically delete head branches` feature in your repository settings.
Reference: https://github.com/peter-evans/create-pull-request?tab=readme-ov-file#delete-branch

lirenjie95 commented:
> @lirenjie95 I also added a small clarification on translation stewardship. As above, stewardship is more about reviewing or helping triage - though of course you are welcome to make translations too (as a contributor)! Let me know what your thoughts are about this approach to stewardship on DevOps.

Thank you for the update and clarification. It's an honor to serve as a steward to p5.js. I would like to review PRs and assist new contributors in understanding the bug fixing process in both translation and building process.

VANSH3104 commented:
@ksen0 Thank you for the recognition! I'd be honored to serve as a steward for Documentation. I look forward to helping review PRs, guide contributors, and improve the stewardship process.

error-four-o-four commented:
@ksen0 I am truly grateful and appreciate the opportunity. I'm a little unsure whether I can devote enough time but I'll try my best

takshittt commented:
Thank you for the invitation — I’d be happy to take on the role of steward for i18n.
Looking forward to contributing more in this area!

clairep94 commented:
Hi, thanks so much for the invite, I would love to be a steward for p5.js-website :)

lirenjie95 commented on utils/stewards-table.js:
```diff
@@ -0,0 +1,53 @@
+        maintainers.add(user);
+      }
+      areaMap[role] = areaMap[role] || new Set();
+      areaMap[role].add(`@${user}`);
```
Adding `@` here will cause the table content contains `@@user` because we also add `@` at line 49. Please remove it.

ksen0 replied:
Thanks for catching that! I added links in the table at the last minute, and the typo stayed.

lirenjie95 commented on utils/stewards-table.js:
```diff
@@ -0,0 +1,53 @@
+          if (main === 'i18n' && !supportedi18n.has(sub)) return;
+          const key = `${main} (${sub})`;
+          areaMap[key] = areaMap[key] || new Set();
+          areaMap[key].add(`@${user}`);
```
Same as line 18.

lirenjie95 commented on utils/stewards-table.js:
```diff
@@ -0,0 +1,53 @@
+let readme = fs.readFileSync('README.md', 'utf8');
+
+readme = readme.replace(
+  /\| *Area *\|.*\n\|[-| ]+\|\n(?:\|.*\|\n?)*/g,
```
There's a better regex to match the stewards table: `/\| *Area[\s\S]*?\| *Steward\(s\) *\|[\r\n]+\|[\s\-|]+\|[\r\n]+(?:\|.*\|[\r\n]+)+/`

ksen0 commented on contributor_docs/steward_guidelines.md:
```diff
@@ -28,6 +32,57 @@ Whether you have just joined us as a steward, are a seasoned maintainer of p5.js
+- **Math**: Includes both external Math API and internal performance improvements
+- **Shapes**: Includes working with custom shapes in both 1.11.x and 2.x versions of p5.js
+- **Maintainers**: This group can merge PRs
+- **p5.sound.js**: The [new p5.sound.js](https://github.com/processing/p5.sound.js) addon library
```
It seems it's made its way into the dictionary of my mind. Add-on was definitely the intent!
